File: TourApp/data/src/main/java/io/github/pedrofraca/data/datasource/model/Resource.java
package io.github.pedrofraca.data.datasource.model;

import androidx.annotation.Nullable;

public class Resource<T> {

    private Status mStatus;
    private T mData;
    private Throwable error;

    public Resource(Status status, T data) {
        mStatus = status;
        this.mData = data;
    }

    @Nullable
    public <T> T getData() {
        return (T) mData;
    }

    public Status getStatus(){
        return mStatus;
    }

    @Nullable
    public Throwable getError() {
        return error;
    }

    public void setError(Throwable error) {
        this.error = error;
    }

    public enum Status {
        ERROR,
        OK,
        CACHE
    }
}

File: TourApp/data/src/main/java/io/github/pedrofraca/data/datasource/model/Resource.kt
package io.github.pedrofraca.data.datasource.model

class Resource<T>(val status: Status, private val mData: T) {
    var error: Throwable? = null
    fun <T> getData(): T? {
        return mData as T
    }

    enum class Status {
        ERROR, OK, CACHE
    }

}